go/vt/vttablet/tabletserver/vstreamer/main_test.go

```diff
@@ -0,0 +1,368 @@
+		// circular dependencies
+		engine = NewEngine(env.TabletEnv, env.SrvTopo, env.SchemaEngine, nil, env.Cells[0])
+		engine.InitDBConfig(env.KeyspaceName, env.ShardName)
+		engine.Open()
+		defer engine.Close()
+
+		return m.Run()
+	}()
+	os.Exit(exitCode)
+}
+
+func newEngine(t *testing.T, ctx context.Context, binlogRowImage string) {
+	if engine != nil {
+		engine.Close()
+	}
+	if env != nil {
+		env.Close()
+	}
+	var err error
+	env, err = testenv.Init(ctx)
+	require.NoError(t, err)
+
+	setBinlogRowImage(t, binlogRowImage)
+
+	// engine cannot be initialized in testenv because it introduces
+	// circular dependencies
+	engine = NewEngine(env.TabletEnv, env.SrvTopo, env.SchemaEngine, nil, env.Cells[0])
+	engine.InitDBConfig(env.KeyspaceName, env.ShardName)
+	engine.Open()
+}
+
+func customEngine(t *testing.T, modifier func(mysql.ConnParams) mysql.ConnParams) *Engine {
+	original, err := env.Dbcfgs.AppWithDB().MysqlParams()
+	require.NoError(t, err)
+	modified := modifier(*original)
+	cfg := env.TabletEnv.Config().Clone()
+	cfg.DB = dbconfigs.NewTestDBConfigs(modified, modified, modified.DbName)
+
+	engine := NewEngine(tabletenv.NewEnv(vtenv.NewTestEnv(), cfg, "VStreamerTest"), env.SrvTopo, env.SchemaEngine, nil, env.Cells[0])
+	engine.InitDBConfig(env.KeyspaceName, env.ShardName)
+	engine.Open()
+	return engine
+}
+
+func setBinlogRowImage(t *testing.T, mode string) {
+	execStatements(t, []string{
+		fmt.Sprintf("set @@binlog_row_image='%s'", mode),
+		fmt.Sprintf("set @@session.binlog_row_image='%s'", mode),
+		fmt.Sprintf("set @@global.binlog_row_image='%s'", mode),
+	})
+}
+
+func runCases(t *testing.T, filter *binlogdatapb.Filter, testcases []testcase, position string, tablePK []*binlogdatapb.TableLastPK) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	wg, ch := startStream(ctx, t, filter, position, tablePK)
+	defer wg.Wait()
+	// If position is 'current', we wait for a heartbeat to be
+	// sure the vstreamer has started.
+	if position == "current" {
+		log.Infof("Starting stream with current position")
+		expectLog(ctx, t, "current pos", ch, [][]string{{`gtid`, `type:OTHER`}})
+	}
+	log.Infof("Starting to run test cases")
+	for _, tcase := range testcases {
+		switch input := tcase.input.(type) {
+		case []string:
+			execStatements(t, input)
+		case string:
+			execStatement(t, input)
+		default:
+			t.Fatalf("unexpected input: %#v", input)
+		}
+		engine.se.Reload(ctx)
+		expectLog(ctx, t, tcase.input, ch, tcase.output)
+	}
+	cancel()
+	if evs, ok := <-ch; ok {
+		t.Fatalf("unexpected evs: %v", evs)
+	}
+	log.Infof("Last line of runCases")
+}
+
+func expectLog(ctx context.Context, t *testing.T, input any, ch <-chan []*binlogdatapb.VEvent, output [][]string) {
+	timer := time.NewTimer(1 * time.Minute)
+	defer timer.Stop()
+	for _, wantset := range output {
+		var evs []*binlogdatapb.VEvent
+		for {
+			select {
+			case allevs, ok := <-ch:
+				if !ok {
+					t.Fatal("expectLog: not ok, stream ended early")
```

Comment:
Is there any reason to use `t.Fatal()` anymore rather than e.g. `requireTrue(t, ok, "expectLog: not ok, stream ended early")` or `require.FailNow("expectLog: not ok, stream ended early")`? I don't think so, but I also know that we're just moving this code so up to you if you want to change it or not. We can always do it later.